### DISCUSSION TITLE
Gentle suggestion about the new plots implementation

### DISCUSSION THREAD

**Discussion by DenisGoryunov** [pwwang/vcfstats#20] [Q&A] [answered]
The vcfstats are able to produce a plenty of informative plots to explore and access genotypes in the vcf file. I'd like to list several currently not supported by vcfstats plotting possibilities which are in my opinion would help in understanding of the content of the vcf file.
1. Barplot with just two bars - one for transitions and one for transversions respectively. Calculate the transitions/transversions ratio and draw it somewhere nearby the barplot.

2. Plot the DP value from the INFO field of vcf. It may be a histogram showing the DP distribution across the whole genome or one plot with the multiple boxplots on it (one boxplot for each chromosome).

3. Plot the per sample percent of the missing genotype calls (single histogram for all samples).

4. Plot the per locus percent of the missing genotype calls ( single histogram for the all loci in genome or plot with multiple boxplots - one per chromosome).

5. Barplot reflecting the percent of positions containing only the homozygous (only HOM REF + only HOM ALT + positions where both the HOM REF and HOM ALT are observed) calls to the total variants number on the chromosome (number of bars is equal to the chromosome number).

Regards,
Denis

**Comment by pwwang** [comment, marked as answer]
Note that you need to upgrade `vcfstats` to `v0.3.0`:

1. TiTv numbers
    ```shell
    vcfstats --vcf examples/sample.vcf \
    --outdir examples/ \
    --config examples/config.toml \
    --title 'TiTv numbers' --figtype col \
    --formula 'COUNT(1) ~ TITV' \
    --ggs 'theme(axis_text_x=element_text(vjust=1))'
    ```

    ![titv-numbers col](https://user-images.githubusercontent.com/1188067/152403158-7f4d5a70-a242-482f-800d-7129b6acebe4.png)

2. INFO_DP
    ```shell
    vcfstats --vcf examples/sample.vcf \        
    --outdir examples/ \
    --config examples/config.toml \
    --title "INFO_DP" \
    --macro examples/mymacros.py \
    --formula 'INFO_DP ~ 1' 
    ```
    
    ![info-dp histogram](https://user-images.githubusercontent.com/1188067/152404102-e43ee8ee-54b7-468c-8d71-133d8b8953a5.png)
  
    ```shell
    vcfstats --vcf examples/sample.vcf \    
    --outdir examples/ \
    --config examples/config.toml \
    --title "INFO_DP-Chroms" \
    --macro examples/mymacros.py \
    --formula 'INFO_DP ~ CONTIG' \
    --figtype boxplot
    ```
    ![info-dp-chroms boxplot](https://user-images.githubusercontent.com/1188067/152404279-088c1153-dbf7-4c7d-8c6d-d8666ac9521a.png)


3. Number of missing calls
    ```shell
    vcfstats --vcf examples/sample.vcf \   
    --outdir examples/ \
    --config examples/config.toml \
    --title "N_missings" \
    --macro examples/mymacros.py \
    --formula 'SUM(MISSINGs) ~ SAMPLES'
    ```
    ![n-missings col](https://user-images.githubusercontent.com/1188067/152420291-8afc1af7-8eab-41a9-817d-db5d3b911b48.png)

4. Per locus missings
    ```shell
    vcfstats --vcf examples/sample.vcf \        
    --outdir examples/ \
    --config examples/config.toml \
    --title "Per_locus_missings" \
    --macro examples/mymacros.py \
    --formula 'N_MISSING ~ 1'
    ```
    ![per-locus-missings histogram](https://user-images.githubusercontent.com/1188067/152420941-6e096583-10ce-43b4-b446-8901232dc7d3.png)

    ```shell
    vcfstats --vcf examples/sample.vcf \        
    --outdir examples/ \
    --config examples/config.toml \
    --title "Per_locus_missings" \
    --macro examples/mymacros.py \
    --formula 'N_MISSING ~ CONTIG' \
    --figtype boxplot
    ```
    ![per-locus-missings boxplot](https://user-images.githubusercontent.com/1188067/152421167-bb948689-7133-4e75-86d0-c8fb67d07167.png)

5. N_hets

    ```shell
    vcfstats --vcf examples/sample.vcf \               
    --outdir examples/ \
    --config examples/config.toml \
    --title "N_hets-sample1" \
    --formula 'COUNT(1, GTTYPEs[HET]{0}) ~ CONTIG'
    ```
    
    ![n-hets-sample1 col](https://user-images.githubusercontent.com/1188067/152421968-c425d69d-69a4-425a-b9e7-5fa0212cf1f8.png)

**Comment by DenisGoryunov** [comment]
Hi,
Thank you  so much! 
I'll try it on my data.
I'm wondering, if it would be possible to plot the percent of the heterozygous calls per locus as a histogram and multiple boxplots (one for each chromosome as on the "INFO_DP-Chroms" ) in addition to the figure "5.N_hets"?
Thank you again.

**Comment by pwwang** [comment]
#### % Het calls distribution:

```shell
vcfstats --vcf examples/sample.vcf \
    --outdir examples/ \  
    --config examples/config.toml \
    --title "Percent_hets" \
    --formula 'Percent_HETs ~ 1' \
    --macro examples/mymacros.py
```
![percent-hets histogram](https://user-images.githubusercontent.com/1188067/152446100-268a9967-06d6-47f0-83de-a92e7444ab2e.png)

#### % Het calls distribution against chromosomes:
```shell
vcfstats --vcf examples/sample.vcf \      
      --outdir examples/ \
      --config examples/config.toml \
      --title "Percent_hets_chroms" \
      --formula 'Percent_HETs ~ CONTIG' \
      --macro examples/mymacros.py \
      --figtype boxplot
```
![percent-hets-chroms boxplot](https://user-images.githubusercontent.com/1188067/152446292-c8ddf3c7-4f0f-4c45-990f-de83ba1f34ec.png)

**Comment by DenisGoryunov** [comment]
Hi,
Thank you! The plots "Percent_hets" and "Percent_hets_chroms" doesn't work. I got errors:
"Traceback (most recent call last):
  File "/home/tkiy/miniconda3/envs/VCF_stat/bin/vcfstats", line 8, in <module>
    sys.exit(main())
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/cli.py", line 195, in main
    ones = get_instances(opts, samples)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/cli.py", line 91, in get_instances
    Instance(
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/instance.py", line 142, in __init__
    self.formula = Formula(formula, samples, passed, title)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 280, in __init__
    self.Y, self.X = PARSER.parse(formula)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/lark.py", line 561, in parse
    return self.parser.parse(text, start=start, on_error=on_error)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parser_frontends.py", line 107, in parse
    return self.parser.parse(stream, start, **kw)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 41, in parse
    return self.parser.parse(lexer, start)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 171, in parse
    return self.parse_from_state(parser_state)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 179, in parse_from_state
    state.feed_token(token)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 150, in feed_token
    value = callbacks[rule](s)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parse_tree_builder.py", line 111, in __call__
    return self.node_builder(filtered)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parse_tree_builder.py", line 309, in f
    return wrapper(func, name, children, None)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/visitors.py", line 390, in _vargs_inline
    return f(*children)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/visitors.py", line 374, in f
    return _f(self, *args, **kwargs)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 20, in term
    return Term(str(name), items, samples)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 72, in __init__
    raise ValueError("Term {!r} has not been registered.".format(name))
ValueError: Term 'Percent_HETs' has not been registered.
Exception ignored in: <function Instance.__del__ at 0x7febe2fae560>
Traceback (most recent call last):
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/instance.py", line 176, in __del__
    del self.data
AttributeError: data
"

"Traceback (most recent call last):
  File "/home/tkiy/miniconda3/envs/VCF_stat/bin/vcfstats", line 8, in <module>
    sys.exit(main())
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/cli.py", line 195, in main
    ones = get_instances(opts, samples)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/cli.py", line 91, in get_instances
    Instance(
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/instance.py", line 142, in __init__
    self.formula = Formula(formula, samples, passed, title)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 280, in __init__
    self.Y, self.X = PARSER.parse(formula)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/lark.py", line 561, in parse
    return self.parser.parse(text, start=start, on_error=on_error)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parser_frontends.py", line 107, in parse
    return self.parser.parse(stream, start, **kw)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 41, in parse
    return self.parser.parse(lexer, start)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 171, in parse
    return self.parse_from_state(parser_state)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 179, in parse_from_state
    state.feed_token(token)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 150, in feed_token
    value = callbacks[rule](s)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parse_tree_builder.py", line 111, in __call__
    return self.node_builder(filtered)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parse_tree_builder.py", line 309, in f
    return wrapper(func, name, children, None)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/visitors.py", line 390, in _vargs_inline
    return f(*children)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/visitors.py", line 374, in f
    return _f(self, *args, **kwargs)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 20, in term
    return Term(str(name), items, samples)
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 72, in __init__
    raise ValueError("Term {!r} has not been registered.".format(name))
ValueError: Term 'Percent_HETs' has not been registered.
Exception ignored in: <function Instance.__del__ at 0x7fcd4e4c6680>
Traceback (most recent call last):
  File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/instance.py", line 176, in __del__
    del self.data
AttributeError: data
"

**Comment by DenisGoryunov** [comment]
Additionally, there are a two warnings for the "INFO_DP" and "Per_locus_missings":

"INFO_DP: (plotnine) /home/tkiy/miniconda3/envs
                             /VCF_stat/lib/python3.10/site-packages/plotnine/sta
                             ts/stat_bin.py:95: PlotnineWarning: 'stat_bin()'
                             using 'bins = 2546'. Pick better value with
                             'binwidth'.
"

"Per_locus_missings: (plotnine) /home/tkiy/mini
                             conda3/envs/VCF_stat/lib/python3.10/site-packages/p
                             lotnine/stats/stat_bin.py:95: PlotnineWarning:
                             'stat_bin()' using 'bins = 270'. Pick better value
                             with 'binwidth'.
"
The INFO_DP plot looks so tiny regarding the scale of the X axis. How can i manage that?
If you will need the examples of the plots, could you explain me please, how can i attach it here?

**Comment by pwwang** [comment]
> Hi, Thank you! The plots "Percent_hets" and "Percent_hets_chroms" doesn't work. I got errors: "Traceback (most recent call last): File "/home/tkiy/miniconda3/envs/VCF_stat/bin/vcfstats", line 8, in sys.exit(main()) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/cli.py", line 195, in main ones = get_instances(opts, samples) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/cli.py", line 91, in get_instances Instance( File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/instance.py", line 142, in **init** self.formula = Formula(formula, samples, passed, title) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 280, in **init** self.Y, self.X = PARSER.parse(formula) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/lark.py", line 561, in parse return self.parser.parse(text, start=start, on_error=on_error) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parser_frontends.py", line 107, in parse return self.parser.parse(stream, start, **kw) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 41, in parse return self.parser.parse(lexer, start) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 171, in parse return self.parse_from_state(parser_state) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 179, in parse_from_state state.feed_token(token) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 150, in feed_token value = callbacks[rule](s) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parse_tree_builder.py", line 111, in **call** return self.node_builder(filtered) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parse_tree_builder.py", line 309, in f return wrapper(func, name, children, None) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/visitors.py", line 390, in _vargs_inline return f(*children) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/visitors.py", line 374, in f return _f(self, *args, **kwargs) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 20, in term return Term(str(name), items, samples) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 72, in **init** raise ValueError("Term {!r} has not been registered.".format(name)) ValueError: Term 'Percent_HETs' has not been registered. Exception ignored in: <function Instance.**del** at 0x7febe2fae560> Traceback (most recent call last): File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/instance.py", line 176, in **del** del self.data AttributeError: data "
> 
> "Traceback (most recent call last): File "/home/tkiy/miniconda3/envs/VCF_stat/bin/vcfstats", line 8, in sys.exit(main()) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/cli.py", line 195, in main ones = get_instances(opts, samples) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/cli.py", line 91, in get_instances Instance( File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/instance.py", line 142, in **init** self.formula = Formula(formula, samples, passed, title) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 280, in **init** self.Y, self.X = PARSER.parse(formula) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/lark.py", line 561, in parse return self.parser.parse(text, start=start, on_error=on_error) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parser_frontends.py", line 107, in parse return self.parser.parse(stream, start, **kw) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 41, in parse return self.parser.parse(lexer, start) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 171, in parse return self.parse_from_state(parser_state) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 179, in parse_from_state state.feed_token(token) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parsers/lalr_parser.py", line 150, in feed_token value = callbacks[rule](s) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parse_tree_builder.py", line 111, in **call** return self.node_builder(filtered) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/parse_tree_builder.py", line 309, in f return wrapper(func, name, children, None) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/visitors.py", line 390, in _vargs_inline return f(*children) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/lark/visitors.py", line 374, in f return _f(self, *args, **kwargs) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 20, in term return Term(str(name), items, samples) File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/formula.py", line 72, in **init** raise ValueError("Term {!r} has not been registered.".format(name)) ValueError: Term 'Percent_HETs' has not been registered. Exception ignored in: <function Instance.**del** at 0x7fcd4e4c6680> Traceback (most recent call last): File "/home/tkiy/miniconda3/envs/VCF_stat/lib/python3.10/site-packages/vcfstats/instance.py", line 176, in **del** del self.data AttributeError: data "

It says `Percent_HETs` macro cannot be found, have you tried the `examples/mymacros.py` file with `--macro` argument?

**Comment by pwwang** [comment]
> The INFO_DP plot looks so tiny regarding the scale of the X axis. How can i manage that?
> If you will need the examples of the plots, could you explain me please, how can i attach it here?

There is no way to interpolate the binwidth/nbins for now. You probably want to try `density` or `freqpoly` plots.

**Comment by DenisGoryunov** [comment]
Hi,
I've checked the command lines i used for the "Percent_hets" and "Percent_hets_chroms" plots. I used `--macro` for these plots. The rest plots using `--macro` argument works fine. May be you not pushed the code for these two plots as they ware the latest changes you made? How can i check that?

**Comment by DenisGoryunov** [comment]
There are no decorators for "Percent_hets" and "Percent_hets_chroms" plots in the `mymacros.py` file:

```
from vcfstats.macros import continuous


@continuous
def INFO_DP(variant):
    """DP from INFO"""
    return variant.INFO["DP"]


@continuous
def MISSINGs(variant):
    """DP from INFO"""
    # convert boolean array to int
    # gts012 = True
    return (variant.gt_types == 3) + 0


@continuous
def N_MISSING(variant):
    """DP from INFO"""
    # convert boolean array to int
    # gts012 = True
    return variant.num_unknown

```

**Comment by DenisGoryunov** [comment]
I've tried both `density` and `freqpoly` options for the `INFO_DP` plot. It still looks weird:
![info-dp freqpoly-1](https://user-images.githubusercontent.com/10725817/153016525-2a89047a-af27-4844-8eb3-ec115eaf06a4.png)
![info-dp density](https://user-images.githubusercontent.com/10725817/153016680-cd115518-6755-4b5c-8757-5327c82d0bea.png)
It is required approximately 90Gb of RAM to produce The `INFO_DP` plot on my vcf file (110K SNPs for 96 individuals).
The plot `INFO_DP-Chroms` is weird too:
![info-dp-chroms boxplot](https://user-images.githubusercontent.com/10725817/153017279-cb066cfc-03c0-4010-8556-c7674e4ce355.png)

May be  all that because i deal with the GBS data.

**Comment by pwwang** [comment]
> Hi, I've checked the command lines i used for the "Percent_hets" and "Percent_hets_chroms" plots. I used `--macro` for these plots. The rest plots using `--macro` argument works fine. May be you not pushed the code for these two plots as they ware the latest changes you made? How can i check that?

I forgot to merge the commit from `dev` branch. Now it should be there.

**Comment by pwwang** [comment]
> I've tried both density and freqpoly options for the INFO_DP plot. It still looks weird:

That could also be due to the skewness of your depth distribution. 
You can also try `--savedata` to see what the data for plotting looks like.

**Comment by DenisGoryunov** [comment]
Hi,
Thank you for your response. The `Percent_hets` and `Percent_hets_chroms` plots are working now. Unfortunately these are not very informative for my data. Is it possible with `--savedata` argument to export the statistics (e.g. `Percent_hets_chroms` for the whole data set ) to the file to import and visualize it in the` R` environment later?
![proportion-hets-chroms boxplot](https://user-images.githubusercontent.com/10725817/153258957-3dae4f52-71b0-4fb2-8135-d48cec583d3d.png)


![proportion-hets histogram](https://user-images.githubusercontent.com/10725817/153259170-a2174a04-cbdd-4724-8826-63f466b5210a.png)

**Comment by pwwang** [comment]
Sure you can. To save that for each locus, just use the formula `Percent_HETs ~ 1` with `--savedata`.
You may also try to add filters for your plots using `vcfstats`. For example: `Percent_HETs[0, 0.1] ~ CONTIG` to plot loci with `Percent_HETs` between 0 and 0.1 only.

**Comment by DenisGoryunov** [comment]
Thank you so much! These options are extremely helpful!

**Comment by DenisGoryunov** [comment]
Hi,
I'm wondering how can i cite `vcfstats` in the publications?
Regards,
Denis

**Comment by pwwang** [comment]
You can just link the url of the repo. It is not published yet.

**Comment by DenisGoryunov** [comment]
Thank you!

**Comment by DenisGoryunov** [comment]
Hi,
I'm wondering if it would be possible to make a two additional plots?

1. Barplot with just the two bars - number of diallelic variants and number of multiallelic variants.
2. More detailed barplot reflecting the number of diallelic, tri-, tetra-, etc. allelic variants.

Regards,
Denis

**Comment by pwwang** [comment]
By "diallelic" I assume you meant "biallelic"?

**Comment by DenisGoryunov** [comment]
Hi,
Yes,  biallelic.

Regards,
Denis

вт, 17 мая 2022 г. в 20:47, pwwang ***@***.***>:

**Reply by pwwang** [reply]
You just need to define two macros:

```python
@categorical
def Allelic_Type(variant):
    """Get allelic type, either biallelic or multiallelic"""
    return "biallelic" if len(variant.ALT) == 1 else "multi-allelic"


@categorical
def N_Allelic(variant):
    """Get number of alleles"""
    return len(variant.ALT) + 1
```

See https://github.com/pwwang/vcfstats/blob/dev/examples/mymacros.py

Then:

```shell
vcfstats --vcf examples/multiallelic.vcf \
    --outdir examples/ \
    --config examples/config.toml \
    --title "Number of biallelics and multiallelics" \
    --macro examples/mymacros.py \
    --figtype col \
    --formula "COUNT(1) ~ Allelic_Type" \
    --ggs "theme(axis_text_x = element_text(angle = 90, vjust = 1, hjust=0.5)); ylab('Count')"
```

![bi/multiallelec](https://github.com/pwwang/vcfstats/raw/dev/examples/number-of-biallelics-and-multiallelics.col.png)

```shell
vcfstats --vcf examples/multiallelic.vcf \
    --outdir examples/ \
    --config examples/config.toml \
    --title "Number of variants with different number of alleles" \
    --macro examples/mymacros.py \
    --figtype col --formula "COUNT(1) ~ N_Allelic" \
    --ggs "ylab('Count')"
```

![n_allelic](https://github.com/pwwang/vcfstats/raw/dev/examples/number-of-variants-with-different-number-of-alleles.col.png)

**Comment by DenisGoryunov** [comment]
Hi,
Thank you for your reply!
Both commands work for me. In my vcf file ( produced on GBS data) there are only biallelic SNPs. However, the plot "Number of variants with different number of alleles" looks a bit weird to me, as there are floats on the X axis. I'd expect to see only integers not floats on the X axis of the plot, perhaps only "2" in case of the biallelic data.
![number-of-variants-with-different-number-of-alleles col](https://user-images.githubusercontent.com/10725817/169360281-8e821ab9-952a-4e2f-b684-f6c82850ae07.png)

Regards,
Denis

**Reply by pwwang** [reply]
How about `--ggs "scale_x_discrete(limits=[2])"`?

**Comment by DenisGoryunov** [comment]
Thanks a lot !
